lib/cpuinfo.c

```diff
@@ -477,6 +491,60 @@ cpuinfo_build_topo(void)
         return l_cpu;
 }
 
+int detect_vendor(void)
+{
+	int ret = 0;
+	struct cpuid_out vendor;
+
+	lcpuid(0x0, 0x0, &vendor);
+	if (vendor.ebx == 0x756e6547 && vendor.edx == 0x49656e69 &&
+	    vendor.ecx == 0x6c65746e) {
+		cpu_vendor = PQOS_VENDOR_INTEL;
+	} else if (vendor.ebx == 0x68747541 && vendor.edx == 0x69746E65 &&
+		   vendor.ecx ==   0x444D4163) {
+		cpu_vendor = PQOS_VENDOR_AMD;
+	} else {
+		ret = -EFAULT;
+	}
+	return ret;
+}
+
+/**
+ * Initialize pointers for the vendors
+ */
+int
+init_functions(struct pqos_vendor_config *ptr)
+{
+        enum pqos_vendor vendor = pqos_get_vendor();
+
+        /**
+         * Make sure to initialize all the pointers to avoid
+         * NULL check while calling
+         */
+        if (vendor == PQOS_VENDOR_INTEL) {
+                ptr->cpuid_cache_leaf = 4;
+                ptr->mba_max = PQOS_MBA_LINEAR_MAX;
+                ptr->mba_msr_reg = PQOS_MSR_MBA_MASK_START;
+                ptr->hw_mba_get = hw_mba_get;
+                ptr->hw_mba_set = hw_mba_set;
+                ptr->os_mba_get = os_mba_get;
+                ptr->os_mba_set = os_mba_set;
```

Comment:
Does not compile on FreeBSD

```
/usr/local/bin/ld: ../lib/libpqos.so: undefined reference to `os_mba_set'
/usr/local/bin/ld: ../lib/libpqos.so: undefined reference to `os_mba_get'
```

```suggestion
#ifdef __linux__
                ptr->os_mba_get = os_mba_get;
                ptr->os_mba_set = os_mba_set;
#endif
```